### PR TITLE
Add support for CoreDNS as DNS provider

### PR DESCRIPTION
**Pull request by etiennetremel**
Add support for CoreDNS as DNS provider. Fix https://github.com/kubernetes-incubator/kube-aws/issues/1075

### CODE REVIEW

File: core/controlplane/config/config.go [hunk 1]
```diff
@@ -817,6 +820,7 @@ type KubeDnsAutoscaler struct {
 }
 
 type KubeDns struct {
```
Comment (mumoshu) [inline core/controlplane/config/config.go]: Yeah, this should retain the name for backward compatibility for now.
But how about renaming this to `ClusterLocalDns` in the **future**?
`KubeDns` sounds like literally specific to `kube-dns`, that is just one of cluster local DNS server implementations available to Kubernetes.

**Comment by c-knowles** [comment]
If a user switches provider on an existing cluster, do both deployments run side by side and one has to be manually deleted? We may want to add a note about that.

File: core/controlplane/config/templates/cloud-config-controller [hunk 2]
```diff
@@ -3642,9 +3726,12 @@ write_files:
                   - /cluster-proportional-autoscaler
                   - --namespace=kube-system
                   - --configmap=kube-dns-autoscaler
+                  {{- if eq .KubeDns.Provider "coredns" }}
+                  - --target=Deployment/coredns
+                  {{- else }}
                   - --target=Deployment/kube-dns
+                  {{- end }}
                   - --default-params={"linear":{"coresPerReplica":{{ .KubeDns.Autoscaler.CoresPerReplica }},"nodesPerReplica":{{ .KubeDns.Autoscaler.NodesPerReplica }},"min":{{ .KubeDns.Autoscaler.Min}}}}
-                  - --logtostderr=true
```
Comment (c-knowles) [inline core/controlplane/config/templates/cloud-config-controller]: Was `logtostderr` intentionally removed here?

Comment (mumoshu) [reply core/controlplane/config/templates/cloud-config-controller]: No, at least for me! Good catch.

**Comment by c-knowles** [comment]
I was trying to figure out if https://github.com/kubernetes-incubator/kube-aws/pull/792 is still really needed when using CoreDNS. There seems to be some discussion in https://github.com/kubernetes/kubernetes/issues/32749 about that, it might be that we could deploy CoreDNS as a Daemon Set and reduce the component count involved.

**Comment by mumoshu** [comment]
@c-knowles Hey!

I got a chance to revisit this. And it seems like the momentum is to use CoreDNS as a node-local dns caching agent, as you've said.

See the sig-network proposal at https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0030-nodelocal-dns-cache.md, which is implemented in the upstream at https://github.com/kubernetes/kubernetes/pull/70555

**Pull request merged**